services/sidecar/rollup/derivation/batch_builder.go

```diff
@@ -81,52 +81,62 @@ func (b *batchBuilder) Build() (*BatchAttributes, error) {
 	if err != nil {
 		return nil, fmt.Errorf("failed to build batch: %w", err)
 	}
-	b.builtBatchAttrs = batchAttrs
-	return batchAttrs, nil
+	batchData, err := encodeBatch(batchAttrs)
+	if err != nil {
+		return nil, fmt.Errorf("failed to encode batch: %w", err)
+	}
+
+	b.builtBatchData = &batchData
+	return b.builtBatchData, nil
 }
 
 func (b *batchBuilder) Advance() {
-	b.builtBatchAttrs = nil
+	b.builtBatchData = nil
 }
 
 func (b *batchBuilder) serializeToAttrs() (*BatchAttributes, error) {
 	var (
-		contexts, txLengths []*big.Int
-		buf                 = new(bytes.Buffer)
-		numBytes            uint64
-		block               DerivationBlock
-		idx                 int
+		block  DerivationBlock
+		idx    int
+		blocks []BlockAttributes
 	)
+	firstL2BlockNumber := big.NewInt(0).SetUint64(b.pendingBlocks[0].BlockNumber())
+	numBytes := uint64(rlp.IntSize(firstL2BlockNumber.Uint64()))
+
+	// Iterate block-by-block to enforce soft cap on batch size.
 	for idx, block = range b.pendingBlocks {
-		// Construct context (`contexts` is a flat array of 2-tuples)
-		contexts = append(contexts, big.NewInt(0).SetUint64(block.NumTxs()))
-		contexts = append(contexts, big.NewInt(0).SetUint64(block.Timestamp()))
-		numBytes += 2 * 8
-		// Construct txData.
+		blockTimestamp := big.NewInt(0).SetUint64(block.Timestamp())
+		// Calculate size of txData.
+		txListSize := uint64(0)
 		for _, tx := range block.Txs() {
-			curLen := buf.Len()
-			if _, err := buf.Write(tx); err != nil {
-				return nil, err
-			}
-			txSize := buf.Len() - curLen
-			txLengths = append(txLengths, big.NewInt(int64(txSize)))
-			numBytes += uint64(txSize)
+			txListSize += uint64(len(tx))
 		}
+		blocks = append(blocks, BlockAttributes{blockTimestamp, block.Txs()})
+		numBytes += rlp.ListSize(uint64(rlp.IntSize(blockTimestamp.Uint64())) + rlp.ListSize(txListSize))
+
 		// Enforce soft cap on batch size.
-		if numBytes > b.maxBatchSize {
-			log.Info("Reached max batch size", "numBytes", numBytes, "maxBatchSize", b.maxBatchSize)
+		if rlp.ListSize(numBytes) > b.maxBatchSize {
+			log.Info("Reached max batch size", "numBytes", numBytes, "maxBatchSize", b.maxBatchSize, "numBlocks", len(blocks))
 			break
 		}
 	}
 	// Construct batch attributes.
-	var (
-		firstL2BlockNumber = big.NewInt(0).SetUint64(b.pendingBlocks[0].BlockNumber())
-		txBatchVersion     = big.NewInt(0) // TODO: Find a better place to not hard code this
-		attrs              = &BatchAttributes{contexts, txLengths, firstL2BlockNumber, txBatchVersion, buf.Bytes()}
-	)
+	attrs := &BatchAttributes{firstL2BlockNumber, blocks}
 	log.Info("Serialized l2 blocks", "first", firstL2BlockNumber, "last", b.pendingBlocks[idx].BlockNumber())
```

Comment:
is it possible that `idx` overflows `pendingBlocks` when the previous loop reads all pending blocks?

Reply:
it's not possible, this loop:
```for idx, block = range b.pendingBlocks {```
terminates with `idx` equal to the index of last item in the array, and `block` equal to the last item in the array.

its probably cleaner to rewrite as:

```log.Info("Serialized l2 blocks", "first", firstL2BlockNumber, "last", block.Number())```

Reply:
what about the next few lines where there's a `b.pendingBlocks[idx + 1]`?

Reply:
`b.pendingBlocks[idx + 1]` will correctly truncate the slice to an empty slice if all blocks are processed however:

 when `b.pendingBlocks` is empty, the loop will not execute and `idx` will not be assigned a value. When the line `b.pendingBlocks` = `b.pendingBlocks[idx+1:]` executes, it would panic due to an index out of range error.

this shouldn't happen because we don't call `serializeToAttrs` unless there is at least one pending block:

https://github.com/SpecularL2/specular/blob/d8b2a27b619e1dcbc37c3a535558f122415f7041/services/sidecar/rollup/derivation/batch_builder.go#L77

but i will cleanup this code to make it less likely future changes unknowingly break things.